BloomFilter.md

```diff
@@ -49,87 +44,185 @@ pushdown even in columns with high cardinality and when space is at a premium.
+block.
+
+The `filter_check` operation uses the same method as `filter_insert`
+to select a block to operate on, then uses the least significant 32
+bits of its argument as an argument to `block_check` called on that
+block, returning the result.
+
+In the pseudocode below, the modulus operator is represented with the C
+language's "`%`" operator. The "`>>`" operator is used to denote the
+conversion of an unsigned 64-bit integer to an unsigned 32-bit integer
+containing only the most significant 32 bits, and C's cast operator
+"`(unsigned int32)`" is used to denote the conversion of an unsigned
+64-bit integer to an unsigned 32-bit integer containing only the lest
+significant 32 bits.
+
+```
+void filter_insert(SBBF filter, unsigned int64 x) {
+  block b = filter.getBlock((x >> 32) % filter.numberOfBlocks());
+  block_insert(b, (unsigned int32)x)
 }
+```
 
 ```
+boolean filter_check(SBBF filter, unsigned int64 x) {
+  block b = filter.getBlock((x >> 32) % filter.numberOfBlocks());
+  return block_check(b, (unsigned int32)x)
+}
+```
+
+The use of blocks is from Putze et al.'s [Cache-, Hash- and
+Space-Efficient Bloom
+filters](http://algo2.iti.kit.edu/documents/cacheefficientbloomfilters-jea.pdf)
 
-#### Hash Function
-The function used to hash values in the initial implementation is
-[xxHash](https://cyan4973.github.io/xxHash/), using the function XXH64 with a
-seed of 0 and [following the specification version
+To use an SBBF for values of arbitrary Parquet types, we apply a hash
+function to that value - at the time of writing,
+[xxHash](https://cyan4973.github.io/xxHash/), using the function XXH64
+with a seed of 0 and [following the specification version
 0.1.1](https://github.com/Cyan4973/xxHash/blob/v0.7.0/doc/xxhash_spec.md).
 
-#### Build a Bloom filter
-The fact that exactly eight bits are checked during each lookup means that these filters
-are most space efficient when used with an expected false positive rate of about
-0.5%. This is achieved when there are about 11.54 bits for every distinct value inserted
-into the filter.
+#### Sizing an SBBF
 
-To calculate the size the filter should be for another false positive rate `p`, use the
-following formula. The output is in bits per distinct element:
+The `check` operation in SBBFs can return `true` for an argument that
+was never inserted into the SBBF. These are called "false
+positives". There is not a simple closed-form calculation of this
+probability, but empirically, an SBBF that uses 6 bits of space (or
+6/256 blocks) per unique item inserted will have a false positive
+probability of around 10%.
 
-```c
-c = -8 / log(1 - pow(p, 1.0 / 8))
-```
+For example, consider an SBBF that contains 1024 blocks and has been
+loaded with (256/6) * 1024 = 43691 unique keys. Running `filter_check`
+on 100 items that were never `filter_insert`ed will yield about 10
+`true` responses in expectation.
+
+Using more space will reduce the false positive probability:
```

Comment:
In the calculations for previous proposals, we found that over-filling a bloom filter by 2x caused the false-positive rate to be 10x the desired rate. In other words, over-filling is expensive. But, under-filling is also expensive because it takes much more space.

It would be helpful to have a rough calculation here to get the false-positive probability from a number of blocks and an estimate of the number of unique keys, although above it seems to imply that there isn't one. At least stating that it is better to under-fill than over-fill would be good?

Reply:
Yes, there isn't one. I'd hesitate to say it is "better" to overfill than underfill. They're just different tradeoffs.